### ISSUE TITLE
Question: Collaboration on better comment ranking algorithms?

### ISSUE THREAD

**Issue by fdietze** [LemmyNet/lemmy#4924]
### Question

Hey everyone,

we're a small [open research group](https://social-protocols.org/) currently working on new comment ranking algorithms for discussion trees. Our goal is to make identifying and debunking of misinformation in discussions more effective. Technically we're analyzing voting patterns using bayesian statistics.

Is there interest in the lemmy community to collaborate on that goal?

Everything we do is open-source. An earlier project we worked on was a new ranking metric for the Hacker News frontpage: https://github.com/social-protocols/news#readme

Happy to answer any questions.

**Comment by dessalines** [comment]
Sure, seems interesting. I'm all for bettering or adding additional sorts.

**Comment by fdietze** [comment]
Great! The biggest challenge we're currently facing is evaluation with real-life scenarios. We already have simulations which confirm that our algorithms do what we expect them to do, but that's no guarantee that they'll work well in the wild.

I see two options to evaluate right now:
- Implement the algorithm in lemmy itself and deploy it to an instance with users to test
- Evaluate on offline voting data from past lemmy discussions

The exact data the algorithms needs is a vote-stream which contains:
- post_id
- user_id
- vote (up or down)
- parent_id (the parent post of the post being voted on)

Any idea what's the most practical way forward?

**Comment by dullbananas** [comment]
If it doesn't give the highest rank to recent stuff only, then it should be added as a new sort type, maybe called "best" or "fair"

**Comment by fdietze** [comment]
> If it doesn't give the highest rank to recent stuff only, then it should be added as a new sort type, maybe called "best" or "fair"

Right. Our algorithm doesn't sort by date. The most descriptive name, we've come up with so far is "convincing".

A bit more in depth: It empirically measures "convincingness" in voting patterns and bubbles up the most convincing comments for every parent. We define convincing as a comment, which measurably changes voting behavior of its parent. The idea is to focus more attention on convincing comments and their convincing replies recursively. Since misinformation and debunking information is usually convincing, it should (in theory) help to debunk misinformation faster.

**Comment by fdietze** [comment]
I looked a bit into the lemmy database schema. I think the easiest first step is to analyze some existing voting data from comment trees. That's much easier than implementing the full algorithm in this code base. I prepared some queries which return exactly the data we need for a first analysis:

```sql
-- all posts
select id, name, url, body, published from post
```


```sql
-- all comments with their parent_id
SELECT 
    id, 
    post_id, 
    content,
    published,     
    CASE 
        WHEN nlevel(path) = 2 THEN NULL 
        ELSE ltree2text(subpath(path, nlevel(path) - 2, 1)) 
    END AS parent_id
FROM 
    comment
```

```sql
-- anonymized post likes, CHANGE THE SALT!
-- salt must be the same in all queries.
select sha224(('random salt 123' || person_id)::bytea) as person_id, post_id, score, published from post_like;
```

```sql
-- anonymized comment likes, CHANGE THE SALT!
-- salt must be the same in all queries.
select sha224(('random salt 123' || person_id)::bytea) as person_id, comment_id, score, published from comment_like
```

Would it be ok to run those on a popular instance and publish the data (or hand it over in private)? I would then analyse it with our group and share insights here.

**Comment by MV-GH** [comment]
Technically all this data is publicly available through the REST API of any instance. So you can retrieve it that way too, but I would suggest asking permission from an admin.

**Comment by fdietze** [comment]
> Technically all this data is publicly available through the REST API of any instance. So you can retrieve it that way too, but I would suggest asking permission from an admin. 

I just double checked the API docs again. As far as I understand, the API only offers vote aggregates, not individual votes per person. Am I overlooking something?

**Comment by Nothing4You** [comment]
The API does indeed only return individual votes to admins or community moderators, but you can see the number of upvotes and the number of downvotes. Note that typically (some exceptions apply) all posts and comments automatically receive an upvote by their creator.

Does it matter which user a vote is from for the result or is that just due to the way it's implemented without affecting the outcome? If it doesn't affect the outcome it could probably just be substituted with a random id for testing.

It should also be noted that the SQL queries above do not provide anonymization.
They provide pseudonymization, and this is rather easily reversible given the association with post ids and comment ids.
For the majority of posts and comments with just a single upvote this vote will be by the creator, so this is a decent way to de-pseudonymize many users that have posted or commented before.

There are also some other Fediverse applications out there that publicly display votes, which could further be used to almost (remote instances typically don't have 100% coverage of local content) fully de-pseudonymize the entire list of votes.

Considering that Lemmy does not make individual votes public, such a dataset, if provided, should likely not be publicly posted.

**Comment by fdietze** [comment]
> Note that typically (some exceptions apply) all posts and comments automatically receive an upvote by their creator.

Very important point! That allows to deanonymize users.

> Does it matter which user a vote is from for the result or is that just due to the way it's implemented without affecting the outcome? If it doesn't affect the outcome it could probably just be substituted with a random id for testing.

Individual votes matter for the calculation. Imagine a comment A and a replying comment B. We statistically measure if users who upvoted B voted differently on A. That can't be calculated from the counts alone.

But in fact, the author's vote doesn't matter here. It can just be left out of the data.

**Comment by fdietze** [comment]
Thank you for revising the queries! They look much better now. 

Any recommendations for which instance admins I should approch?

**Comment by Nothing4You** [comment]
@dessalines could provide some information about lemmy.ml, which is probably the oldest instance around, although not necessarily the most complete one. due to bugs, limited federation, age or lack of community subscriptions there won't be any single instance that has all data.

lemmy.world as the largest instance is probably worth reaching out to as well.
hexbear.net seems to have the most posts and comments, and it also exists for a lot longer than most others.
sh.itjust.works and lemm.ee are likely also good options with instances that have a lot of data.

https://lemmyverse.net/?order=posts and https://lemmyverse.net/?order=comments might be useful.

you can ignore lemmit.online, that's a reddit repost instance. there are a few more that may be mostly or only mirroring reddit content as well.

also keep in mind that post/comment/user ids are local to an instance, so you can't merge data from multiple instances this way.
if data should be mergable it'd require using the `ap_id`s of posts/comments and `actor_id` of users, plus sharing the salt across those instances.

**Comment by fdietze** [comment]
Do I understand correctly that the `ap_id`/`actor_id` can be used (instead of the `id`) as a global id, so that datasets from multiple instances can be trivially merged? (assuming the same salt)

With that understanding, I revised the queries another time:
- use `ap_id`/`actor_id` instead of `id` in all queries

@Nothing4You if you approve, I'll start approaching instance admins. Thank you for your help!

```sql
-- get all posts
select ap_id, post.name, post.url, post.body, post.published
from post
join community on post.community_id = community.id
where post.deleted = false
and post.removed = false
and community.removed = false
and community.deleted = false
;

-- get all comments
select
    comment.ap_id,
    post.ap_id as post_ap_id,
    comment.content,
    comment.published,
    parent_comment.ap_id as parent_ap_id
from comment
join post on post.id = comment.post_id
join community on post.community_id = community.id
join comment as parent_comment on parent_comment.id = (
    case
        when nlevel(comment.path) = 2 then null
        else ltree2text(subpath(comment.path, nlevel(comment.path) - 2, 1))
    end)::int
where comment.deleted = false
and comment.removed = false
and post.deleted = false
and post.removed = false
and community.removed = false
and community.deleted = false
;


-- pseudonymized post likes, CHANGE THE SALT!
-- salt must be the same in all queries.
select
    sha224(('random salt 123' || person.actor_id)::bytea) as actor_id,
    post.ap_id as post_ap_id,
    post_like.score,
    post_like.published
from post_like
join post on post.id = post_like.post_id
join community on post.community_id = community.id
join person on person.id = person_id
where post_like.person_id != post.creator_id
and post_like.score != 0
and post.deleted = false
and post.removed = false
and community.removed = false
and community.deleted = false
;

-- pseudonymized comment likes, CHANGE THE SALT!
-- salt must be the same in all queries.
select
    sha224(('random salt 123' || person.actor_id)::bytea) as actor_id,
    comment.ap_id as comment_ap_id,
    comment_like.score,
    comment_like.published
from comment_like
join comment on comment.id = comment_like.post_id
join post on post.id = comment.post_id
join community on post.community_id = community.id
join person on person.id = person_id
where comment_like.person_id != comment.creator_id
and comment_like.score != 0
and comment.deleted = false
and comment.removed = false
and post.deleted = false
and post.removed = false
and community.removed = false
and community.deleted = false
;
```

**Comment by Nothing4You** [comment]
> Do I understand correctly that the `ap_id`/`actor_id` can be used (instead of the `id`) as a global id, so that datasets from multiple instances can be trivially merged? (assuming the same salt)

this is correct. those ids will be globally unique identifiers typically. there can be exceptions to this if an instance is torn down and recreated without adjusting the auto increment counters, which can otherwise lead to reuse of identifiers. similarly, user accounts could have been purged at some point, which could be registered again afterwards. the combination of the `ap_id`/`actor_id` and the `published` field of the same object should be good enough to detect accidental false positives, assuming nobody intentionally reused ids and backdated them. i don't have any idea how common such cases of reuse are, but i've seen them in the past occasionally when people were resetting their instance following some issues they were seeing before.

**Comment by dullbananas** [comment]
@fdietze is there now an implementation of this ranking algorithm that can be added to Lemmy?

**Comment by fdietze** [comment]
@dullbananas unfortunately no. We ran out of money and had to give up the research. We never succeeded to raise any money at all.

We also never got our hands on lemmy data to run experiments.

The most important result of our research is a new definition for misinformation. This definition is actionable and grounded in Bayesian statistics and information theory:

```
Misinformation is something you wouldn't believe if you had access to all available information.
```

 If anybody is interested in continuing research in this direction, I'm happy to talk and transfer some insights and ideas.
